ethereum/prices/insert_prices_from_dex_data.sql

```diff
@@ -0,0 +1,200 @@
+CREATE OR REPLACE FUNCTION prices.insert_prices_from_dex_data(start_time timestamptz, end_time timestamptz=now()) RETURNS integer
+LANGUAGE plpgsql AS $function$
+DECLARE r integer;
+BEGIN
+
+WITH trades_with_usd_amount AS (
+    SELECT
+        token_a_address as contract_address,
+        token_a_symbol as symbol,
+        decimals,
+        usd_amount/(token_a_amount_raw/10^decimals) AS price,
+        block_time
+    FROM dex.trades
+    INNER JOIN erc20.tokens ON contract_address = token_a_address
+    WHERE usd_amount  > 0
+    AND category = 'DEX'
+    AND token_a_amount_raw > 0
+    AND block_time >= start_time
+    AND block_time < end_time
+
+    UNION ALL
+
+    SELECT
+        token_b_address as contract_address,
+        token_b_symbol as symbol,
+        decimals,
+        usd_amount/(token_b_amount_raw/10^decimals) AS price,
+        block_time
+    FROM dex.trades
+    INNER JOIN erc20.tokens ON contract_address = token_b_address
+    WHERE usd_amount  > 0
+    AND category = 'DEX'
+    AND token_b_amount_raw > 0
+    AND block_time >= start_time
+    AND block_time < end_time
+),
+grouped_by_hour AS (
+    SELECT
+        date_trunc('hour', block_time) as hour,
+        contract_address,
+        symbol,
+        decimals,
+        (PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY price)) AS median_price,
+        count(1) AS sample_size
+    FROM trades_with_usd_amount
+    GROUP BY 1, 2, 3, 4
+    -- Get previous entries from the table
+    -- This is necessary to provide price continuity for tokens with low swap volume
+    -- when performing historic backfill.
+    UNION ALL
+    SELECT
+        hour,
+        contract_address,
+        symbol,
+        decimals,
+        median_price,
+        sample_size
+    FROM prices.prices_from_dex_data
+    WHERE hour = (select start_ts - INTERVAL '1 hour')
+),
+-- The SQL code in `leaddata`, `generate_hours` and `add_data_for_all_hours`
+-- sets the median_price to the price of the previous hour in case that there
+-- are no swaps in an hour for a token.
+leaddata as 
+(
+    SELECT
+        hour,
+        contract_address,
+        symbol,
+        decimals,
+        median_price,
+        sample_size,
+        lead(hour, 1, now() ) OVER (PARTITION BY contract_address ORDER BY hour asc) AS next_hour
+    FROM grouped_by_hour
+    WHERE sample_size > 0
+),
+generate_hours AS
+(
+    SELECT hour from generate_series(start_time, (select end_time - INTERVAL '1 hour'), '1 hour') g(hour)
+),
+add_data_for_all_hours as 
+(
+    SELECT
+    gen.hour as hour,
+    contract_address,
+    symbol,
+    decimals,
+    median_price,
+    CASE WHEN gen.hour = data.hour THEN sample_size ELSE 0 END AS sample_size
+    FROM leaddata data
+    INNER JOIN generate_hours gen ON data.hour <= gen.hour
+    AND gen.hour < data.next_hour -- Yields an observation for every hour after the first transfer until the next hour with transfer
```

Comment:
This 1 hour now is more than the 10 minute window that was used in `dex.view_token_prices`. Do we expect that this window is wide enough so not to result in missing data from `prices.usd`?

Reply:
I expect this to be fine given that the script starts running at minute 14 of the next hour. Also, every run of the script goes back three days in time and performs retroactive updates where applicable. If there is more complete price data in `dex.trades` since the previous run, the rows in question will be updated.

Reply:
Great! My only question now is wen merge?

Reply:
Merging ongoing. Expect this to be finished in a few hours time. Will run some tests first thing tomorrow.